### PR TITLE
[doc]: Add blog 'Doris analysis: Doris Compaction mechanism analysis' 

### PR DESCRIPTION
**Pull request by ZHbamboo**
# Changes

Add blog 'Doris analysis: Doris Compaction mechanism analysis' in both Chinese and English language with pictures.

Issue Number: #11706

### CODE REVIEW

File: i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+<img alt="图1-1 rowset版本" src="../../../static/images/blogs/doris-compaction-mechanism-parse/Figure_1-1.png"/>
+</div>
+ <p align="center">图1-1 rowset版本</p>
+
+Compaction分为两种类型：base compaction和cumulative compaction。其中cumulative compaction则主要负责将多个最新导入的rowset合并成较大的rowset，而base compaction会将cumulative compaction产生的rowset合入到start version为0的基线数据版本（Base Rowset）中，是一种开销较大的compaction操作。这两种compaction的边界通过cumulative point来确定。base compaction会将cumulative point之前的所有rowset进行合并，cumulative compaction会在cumulative point之后选择相邻的数个rowset进行合并，如图1-2所示。
+
+<div align=center>
+<img alt="图1-2 compaction示意图" src="../../../static/images/blogs/doris-compaction-mechanism-parse/Figure_1-2.png"/>
+</div>
+ <p align="center">图1-2 compaction示意图</p>
```
Comment (hf200012) [inline i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md:53]:
```suggestion
![图1-2 compaction示意图](/images/blogs/doris-compaction-mechanism-parse/Figure_1-2.png")
```

File: i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+# 1 Foreword
+
+Doris的存储引擎通过类似LSM的数据结构提供快速的数据导入支持。对于单一的数据分片（Tablet），新的数据先写入内存结构，随后刷入磁盘，形成一个个不可变更的数据文件，这些数据文件保存在一个rowset中。而Doris的Compaction机制主要负责根据一定的策略对这些Rowset进行合并，将小文件合并成大文件，进而提升查询性能。
+
+每一个rowset都对应一个版本信息，表示当前rowset的版本范围，版本信息中包含了两个字段first和second，first表示当前rowset的起始版本（start version），second表示当前rowset的结束版本（end version），如图1-1所示。每一次数据导入都会生成一个新的数据版本，保存在一个rowset中。未发生过compaction的rowset的版本信息中first字段和second字段相等；执行compaction时，相邻的多个rowset会进行合并，生成一个版本范围更大的rowset，合并生成的rowset的版本信息会包含合并前的所有rowset的版本信息。
+
+<div align=center>
+<img alt="图1-1 rowset版本" src="../../../static/images/blogs/doris-compaction-mechanism-parse/Figure_1-1.png"/>
+</div>
+ <p align="center">图1-1 rowset版本</p>
```
Comment (hf200012) [inline i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md:46]:
```suggestion
![图1-1 rowset版本](/images/blogs/doris-compaction-mechanism-parse/Figure_1-1.png")
```

File: i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+
+# 2 总体架构
+
+## 2.1 “生产者-消费者”模式
+Compaction机制要解决的第一个问题，就是如何选取合适的Tablet进行Compaction。Doris的compaction机制采用“生产者-消费者”（producer-consumer）模式，由producer线程持续生产compaction任务，并将生产出的compaction任务提交给compaction线程池进行消费执行，如图2-1所示。
+
+<div align=center>
+<img alt="图2-1 compaction机制的“生产者-消费者”模式示意图" width="70%" src="../../../static/images/blogs/doris-compaction-mechanism-parse/Figure_2-1.png"/>
+</div>
+ <p align="center">图2-1 compaction机制的“生产者-消费者”模式示意图</p>
```
Comment (hf200012) [inline i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md:63]:
```suggestion
![](/images/blogs/doris-compaction-mechanism-parse/Figure_2-1.png")
```

File: i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+
+Doris BE启动时，会在后台启动一个compaction producer线程，同时会创建一个compaction线程池。producer线程持续地生产compaction任务。
+
+在一轮任务生产过程中，会从每个磁盘各选择出一个tablet执行compaction任务。如果compaction线程池中某个磁盘的compaction任务数达到了上限（通过compaction_task_num_per_disk配置，默认值为2），则这一轮任务生产会跳过该磁盘。如果某一轮生产过程从所有磁盘均没有生产出compaction任务（即compaction线程池中每个磁盘的任务数都已达到上限），则producer线程会进入休眠状态，直到2秒超时唤醒，或线程池中某个compaction任务执行完成被唤醒。
+
+在一轮compaction任务生产过程中，进行单个磁盘的任务生产时，需要遍历BE节点上所有的tablet，首先过滤掉不满足条件的tablet，比如：其他磁盘上的tablet、已经提交给compaction线程池的tablet、正在执行alter操作的tablet、初始化失败的tablet、上次compaction任务失败距离当前时刻的时间间隔小于设定阈值（通过min_compaction_failure_interval_sec配置，默认值为600秒）的tablet，然后从剩余的满足条件的tablet中选择tablet score最高的tablet执行compaction任务。
+
+**tablet score通过如下公式计算：**
+
+<font color=#15A9CA><center><b><i>tablet_score = k1 * scan_frequency + k2 * compaction_score</b></i></center></font>
```
Comment (hf200012) [inline i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md:73]:
```suggestion
`tablet_score = k1 * scan_frequency + k2 * compaction_score`
```

File: i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+## 2.2 permission机制
+
+producer生产出的compaction任务需要提交给compaction线程池执行。为了调节BE节点compaction的内存使用量，Doris增加了对compaction任务提交的permission机制，如图2-2所示。
+
+系统维持一定数量的compaction permits（通过参数total_permits_for_compaction_score配置），每一个compaction任务提交给线程池之前都需要向系统申请permits（permits request），只有获得系统分配的permits之后任务才能被提交给compaction线程池，compaction任务在线程池中执行结束之后需要将自己持有permits归还给系统（permits release）。如果系统当前剩余的可分配的compaction permits数量小于本次compaction任务需要的permits数量，则本次任务提交会被阻塞（compaction任务提交是串行执行的，其他需要提交的任务也会被阻塞），直到有其他compaction任务执行结束并释放permits，使得系统有足够数量的permits分配给当前compaction任务。如果某一个compaction任务需要的permits数量超过系统维持的permits总数，则允许当线程池中所有的任务都执行结束之后，将该compaction任务提交给线程池执行。
+
+<div align=center>
+<img alt="图2-2 compaction任务提交的permission机制" width="60%" src="../../../static/images/blogs/doris-compaction-mechanism-parse/Figure_2-2.png"/>
+</div>
+ <p align="center">图2-2 compaction任务提交的permission机制</p>
```
Comment (hf200012) [inline i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md:92]:
```suggestion
![](/images/blogs/doris-compaction-mechanism-parse/Figure_2-2.png")
```

File: i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+
+Doris中单个compaction任务执行过程中的内存使用量与本次compaction任务合并的segment文件数量有关。一个rowset会包含多个segment文件，而一个compaction任务可能包含多个rowset。因此，使用compaction任务中需要合并的segment文件数量作为compaction任务的permits。通过调整系统维持的compaction permits总量可以对BE节点compaction的内存使用量进行调节。
+
+Compaction任务可以概括为两个阶段：compaction preparation和compaction execution，如图2-3所示。compaction preparation阶段主要是从tablet中选出需要进行版本合并的rowsets，compaction execution阶段主要进行rowsets的版本合并操作。
+
+<div align=center>
+<img alt="图2-3 compaction任务的两阶段划分示意图" width="50%" src="../../../static/images/blogs/doris-compaction-mechanism-parse/Figure_2-3.png"/>
+</div>
+ <p align="center">图2-3 compaction任务的两阶段划分示意图</p>
```
Comment (hf200012) [inline i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md:101]:
```suggestion
![](/images/blogs/doris-compaction-mechanism-parse/Figure_2-3.png")
```

File: i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+
+在Doris中，compaction任务的preparation阶段在permits request之前执行，从tablet中选出需要进行版本合并的rowsets，通过需要合并的segment文件数量计算compaction permits。compaction任务的execution阶段会真正在线程池中执行，进行版本的合并，如图2-4所示。
+
+<div align=center>
+<img alt="图2-4 compaction任务的提交执行示意图" width="70%" src="../../../static/images/blogs/doris-compaction-mechanism-parse/Figure_2-4.png"/>
+</div>
+ <p align="center">图2-4 compaction任务的提交执行示意图</p>
```
Comment (hf200012) [inline i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md:108]:
```suggestion
![](/images/blogs/doris-compaction-mechanism-parse/Figure_2-4.png")
```

File: i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+
+Doris也提供了http接口，支持手动触发单个tablet的cumulative compaction或base compaction。
+
+# 3 Cumulative Compaction
+Doris的cumulative compaction每次会在cumulative point之后选择相邻的数个rowset进行合并，主要包含5个步骤，分别是计算cumulative point、生成candidate rowsets、选择input rowsets、执行rowsets合并以及更新cumulative point，如图3-1。其中，前面三个步骤属于compaction preparation阶段，后面两个步骤属于compaction execution阶段。
+
+<div align=center>
+<img alt="图3-1 cumulative compaction执行流程图" width="80%" src="../../../static/images/blogs/doris-compaction-mechanism-parse/Figure_3-1_cn.png"/>
+</div>
+ <p align="center">图3-1 cumulative compaction执行流程图</p>
```
Comment (hf200012) [inline i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md:120]:
```suggestion
![](/images/blogs/doris-compaction-mechanism-parse/Figure_3-1_cn.png")
```

File: i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+
+## 3.1  计算cumulative point
+
+版本号比cumulative point小的rowset仅会执行base compaction，而版本号比cumulative point大的rowset仅会执行cumulutive compaction。将一个rowset从cumulative 侧移动到base侧（即增大cumulative point）的行为称为一次Promotion。
+
+如果tablet当前的cumulative point值为-1（初始值），则本次计算的cumulative point值不变，仍为-1；否则，执行以下操作进行cumulative point的计算：
+1. 对tablet下所有的rowset按照版本先后进行排序；
+
+2. Doris会通过计算promotion size来决定是否要对一个rowset执行Promotion。根据base rowset（start version为0）的大小计算tablet当前的promotion_size：
+   <font color=#15A9CA><center><b><i>promotion_size = base_rowset_size * ratio</b></i></center></font>
```
Comment (hf200012) [inline i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md:134]:
```suggestion
   `promotion_size = base_rowset_size * ratio`
```

File: i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+   其中，ratio值可以通过cumulative_size_based_promotion_ratio配置，默认值为0.05。promotion_size被限定在cumulative_size_based_promotion_size_mbytes（默认值为1024MB）与cumulative_size_based_promotion_min_size_mbytes（默认值为64MB）之间，promotion size的计算流程如图3-2所示。
+   <div align=center>
+   <img alt="图3-2 promotion size的计算流程图" width="80%" src="../../../static/images/blogs/doris-compaction-mechanism-parse/Figure_3-2_cn.png"/>
+   </div>
+    <p align="center">图3-2 promotion size的计算流程图</p>
```
Comment (hf200012) [inline i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md:139]:
```suggestion
![](/images/blogs/doris-compaction-mechanism-parse/Figure_3-2_cn.png")
```

File: i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+   1. 如果input rowsets中所有的rowset大小之和达到promotion_size，则不需要调整input rowsets。
+   2. 如果存在数据删除版本的记录（last_delete_version 不为-1，即生成input rowsets的遍历过程因为存在删除数据版本而结束），并且input rowsets中的rowset数量不为1，则不需要调整input rowsets；如果存在数据删除版本的记录，并且input rowsets中的rowset数量为1，同时该rowset中的segment文件之间存在overlapping，则不需要调整input rowsets；如果存在数据删除版本的记录，并且input rowsets中的rowset数量为1，同时该rowset中的segment文件之间不存在overlapping，则清空input rowsets。
+   3. 如果不存在数据删除版本的记录（last_delete_version 为-1），则遍历input rowsets中的rowset。从第一个rowset开始，计算当前rowset的大小等级（current_level），同时计算input rowsets中除当前rowset之外的其他rowset大小之和的等级（remain_level），如果current_level > remain_level，则从input rowsets中删除当前rowset，否则，停止遍历。
+
+**【注】** level等级划分由参数cumulative_size_based_promotion_size_mbytes （默认值为1024MB）和cumulative_size_based_compaction_lower_size_mbytes （默认值为64MB）确定。最高的level值为cumulative_size_based_promotion_size_mbytes / 2，下一级level值为上一级level值的1/2，直到level值小于cumulative_size_based_compaction_lower_size_mbytes，则设置该级level值为0，level等级划分流程如图3-3所示。
+
+<div align=center>
+   <img alt="图3-3 level等级划分流程图" width="60%" src="../../../static/images/blogs/doris-compaction-mechanism-parse/Figure_3-3_cn.png"/>
+   </div>
+    <p align="center">图3-3 level等级划分流程图</p>
```
Comment (hf200012) [inline i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md:168]:
```suggestion
![](/images/blogs/doris-compaction-mechanism-parse/Figure_3-3_cn.png")
```

File: i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+在compaction producer线程中，需要依据cumulative compaction score生产cumulative compaction任务。依次遍历tablet中的所有rowset，如果某一个rowset的版本位于cumulative point之后，则将该rowset添加到向量rowset_to_compact。
+1. 如果rowset_to_compact中所有rowset的大小之和超过promotion_size，则rowset_to_compact中所有rowset的score之和为当前tablet的cumulative compaction score，即rowset_to_compact中所有rowset的segment文件数目之和。
+2. 如果rowset_to_compact中所有rowset的大小之和小于promotion_size，则按照版本先后对rowset_to_compact中的rowset进行排序，然后遍历rowset_to_compact中的每一个rowset。计算当前rowset的大小等级（current_level），同时计算rowset_to_compact中除当前rowset之外的其他rowset大小之和的等级（remain_level），如果current_level > remain_level，则从向量rowset_to_compact中删除当前rowset，否则，停止遍历。rowset_to_compact中所有rowset的score之和为当前tablet的cumulative compaction score。
+
+# 4 Base Compaction
+Doris的base compaction会将cumulative point之前的所有rowset进行合并，主要包含3个步骤，分别是选择input rowsets、检查base compaction的执行条件以及执行rowsets合并，如图4-1所示。其中，前面两个步骤属于compaction preparation阶段，最后一个步骤属于compaction execution阶段。
+<div align=center>
+   <img alt="图4-1 base compaction执行流程图" width="60%" src="../../../static/images/blogs/doris-compaction-mechanism-parse/Figure_4-1_cn.png"/>
+   </div>
+    <p align="center">图4-1 base compaction执行流程图</p>
```
Comment (hf200012) [inline i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md:195]:
```suggestion
![](/images/blogs/doris-compaction-mechanism-parse/Figure_4-1_cn.png")
```

File: i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+
+   如果input rowsets中除base rowset（start version为0）之外的其他rowset大小之和与base rowset大小之比超过设定的阈值（通过base_cumulative_delta_ratio配置，默认值为0.3），则当前tablet满足base compaction的执行条件；否则，检查检查下一个条件。
+6. 时间条件
+
+   如果当前tablet上一次成功执行base compaction的时间距离当前时刻超过指定的时间间隔（通过base_compaction_interval_seconds_since_last_operation配置，默认值为86400秒，即1天），则当前tablet满足base compaction的执行条件；否则，当前tablet不满足base compaction的执行条件，本次base compaction任务结束。
+
+<div align=center>
+   <img alt="图4-2 检查base compaction执行条件的流程图" width="60%" src="../../../static/images/blogs/doris-compaction-mechanism-parse/Figure_4-2_cn.png"/>
+   </div>
+    <p align="center">图4-2 检查base compaction执行条件的流程图</p>
```
Comment (hf200012) [inline i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md:225]:
```suggestion
![](/images/blogs/doris-compaction-mechanism-parse/Figure_4-2_cn.png")
```

File: i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+在compaction producer线程中，需要依据base compaction score生产base compaction任务。依次遍历tablet中的每一个rowset，所有位于cumulative point之前的rowset的 score之和为tablet当前的base compaction score，即cumulative point之前的rowset中的segment文件数目之和。
+
+# 5 总结
+本文详细地介绍了Doris的compaction机制。
+
+首先，从producer-consumer模式以及compaction任务提交的permission机制对compaction的总体设计和架构原理进行了剖析；然后，针对cumulative compaction的size_based策略进行了详细地介绍；最后，对base compaction的流程进行了深入地讲解。
+
+Doris通过compaction机制将不同的数据版本进行聚合，将小文件合并成大文件，进而有效地提升了查询性能。
+
+<font color=#15A9CA><b>【TODO】</b></font>
```
Comment (hf200012) [inline i18n/zh-CN/docusaurus-plugin-content-blog/doris-compaction-mechanism-parse.md:240]: remove

**Review by hf200012 — COMMENTED** [review]
Don't embed html tags in markdown

**Comment by luzhijing** [comment]
LGTM

**Pull request closed without merging**